block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -249,11 +248,22 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
   override def accessEquivocationsTracker[A](f: EquivocationsTracker[F] => F[A]): F[A] =
     lock.withPermit(f(KeyValueStoreEquivocationsTracker))
 
-  def addFinalizedBlockHashes(hashes: List[BlockHash]): F[Unit] =
-    blockMetadataIndex.recordFinalized(hashes)
-
-  def recordDirectlyFinalised(blockHash: BlockHash): F[Unit] =
-    blockMetadataIndex.recordDirectlyFinalised(blockHash)
+  /** Record that some hash is directly finalized (detected by finalizer and becomes LFB). */
+  def recordDirectlyFinalized(
+      directlyFinalizedHash: BlockHash,
+      finalizationEffect: Set[BlockHash] => F[Unit]
+  ): F[Unit] =
+    for {
+      dag    <- getRepresentation
```

Comment:
I think it makes sense to have only one method on BlockDagStorage - to record directly finalized hash. And it should compute what else should be finalized indirectly.